packages/runtime/lib/runtime.js

```diff
@@ -1884,45 +1921,155 @@ class Runtime extends EventEmitter {
+        }
+        if (update.workers > MAX_WORKERS) {
+          throw new errors.InvalidArgumentError('workers', `must be less than ${MAX_WORKERS}`)
+        }
+
+        if (currentWorkers === update.workers) {
+          this.logger.warn({ serviceId, workers: update.workers }, 'No change in the number of workers for service')
+          updateWorkers = false
+        } else {
+          updateWorkers = true
+        }
       }
 
-      if (currentWorkers < workers) {
-        await this.#updateWorkerCount(serviceId, workers)
+      let updateHeap
+      if (update.maxHeapTotal !== undefined) {
+        if (typeof update.maxHeapTotal === 'string') {
+          try {
+            update.maxHeapTotal = parseMemorySize(update.maxHeapTotal)
+          } catch {
+            throw new errors.InvalidArgumentError('maxHeapTotal', 'must be a valid memory size')
+          }
+        } else if (typeof update.maxHeapTotal === 'number') {
+          if (update.maxHeapTotal <= 0) {
+            throw new errors.InvalidArgumentError('maxHeapTotal', 'must be greater than 0')
+          }
+        } else {
+          throw new errors.InvalidArgumentError('maxHeapTotal', 'must be a number or a string representing a memory size')
+        }
+
+        if (currentHealth.maxHeapTotal === update.maxHeapTotal) {
+          this.logger.warn({ serviceId, maxHeapTotal: update.maxHeapTotal }, 'No change in the max heap total for service')
+          updateHeap = false
+        } else {
+          updateHeap = true
+        }
+      }
+
+      if (updateWorkers || updateHeap) {
+        validatedUpdates.push({ serviceId, config: serviceConfig, workers: update.workers, maxHeapTotal: update.maxHeapTotal, currentWorkers, currentHealth })
+      }
+    }
+
+    return validatedUpdates
+  }
+
+  async #updateServiceWorkersAndHeap (serviceId, config, serviceConfig, workers, maxHeapTotal, currentWorkers, currentHealth) {
+    if (currentWorkers > workers) {
+      // stop workers
+      await this.#updateServiceWorkers(serviceId, config, serviceConfig, workers, currentWorkers)
+      // update heap for current workers
+      await this.#updateServiceHeap(serviceId, config, serviceConfig, workers, currentHealth, maxHeapTotal)
+    } else {
+      // update service heap
+      await this.#updateServiceConfigHeap(serviceId, maxHeapTotal)
+      // start new workers with new heap
+      await this.#updateServiceWorkers(serviceId, config, serviceConfig, workers, currentWorkers)
+      // update heap for current workers
+      await this.#updateServiceHeap(serviceId, config, serviceConfig, currentWorkers, currentHealth, maxHeapTotal, false)
+    }
+  }
+
+  async #updateServiceHeap (serviceId, config, serviceConfig, currentWorkers, currentHealth, maxHeapTotal, updateConfig = true) {
+    let updated = 0
+    try {
+      if (updateConfig) {
+        await this.#updateServiceConfigHeap(serviceId, maxHeapTotal)
+      }
+
+      for (let i = 0; i < currentWorkers; i++) {
```

Comment:
Add a comment explaining why we are updating them one-by-one instead of in parallel (I think this is correct)